src/equations/serre_green_naghdi_flat_1d.jl

```diff
@@ -0,0 +1,443 @@
+    h, v = q.x
+    dh, dv = dq.x
+
+    @trixi_timeit timer() "hyperbolic terms" begin
+        # Compute all derivatives required below
+        (; h_x, v_x, h2_x, hv_x, v2_x, h2_v_vx_x,
+        h_vx_x, p_x, tmp, M_h, M_h3_3) = cache
+
+        mul!(h_x, D, h)
+        mul!(v_x, D, v)
+        @. tmp = h^2
+        mul!(h2_x, D, tmp)
+        @. tmp = h * v
+        mul!(hv_x, D, tmp)
+        @. tmp = v^2
+        mul!(v2_x, D, tmp)
+
+        @. tmp = h^2 * v * v_x
+        mul!(h2_v_vx_x, D, tmp)
+        @. tmp = h * v_x
+        mul!(h_vx_x, D, tmp)
+        inv6 = 1 / 6
+        @. tmp = (0.5 * h^2 * (h * v_x + h_x * v) * v_x
+                  -
+                  inv6 * h * h2_v_vx_x
+                  -
+                  inv6 * h^2 * v * h_vx_x)
+        mul!(p_x, D, tmp)
+
+        # Plain: h_t + (h v)_x = 0
+        #
+        # Split form for energy conservation:
+        # h_t + h_x v + h v_x = 0
+        @. dh = -(h_x * v + h * v_x)
+
+        # Plain: h v_t + ... = 0
+        #
+        # Split form for energy conservation:
+        @. tmp = -(g * h2_x - g * h * h_x
+                   +
+                   0.5 * h * v2_x
+                   -
+                   0.5 * v^2 * h_x
+                   +
+                   0.5 * hv_x * v
+                   -
+                   0.5 * h * v * v_x
+                   +
+                   p_x)
+    end
+
+    @trixi_timeit timer() "assembling elliptic operator" begin
+        # The code below is equivalent to
+        #   dv .= (Diagonal(h) - Dmat * Diagonal(1/3 .* h.^3) * Dmat) \ tmp
+        # but faster since the symbolic factorization is reused.
+        # Floating point errors accumulate a bit and the system matrix
+        # is not necessarily perfectly symmetric but only up to round-off
+        # errors. We wrap it here to avoid issues with the factorization.
+        @. M_h = h
+        scale_by_mass_matrix!(M_h, D)
+        inv3 = 1 / 3
+        @. M_h3_3 = inv3 * h^3
+        scale_by_mass_matrix!(M_h3_3, D)
+        system_matrix = Symmetric(Diagonal(M_h)
+                                  +
+                                  Dmat' * Diagonal(M_h3_3) * Dmat)
+    end
+
+    @trixi_timeit timer() "solving elliptic system" begin
+        if issparse(system_matrix)
+            (; factorization) = cache
+            cholesky!(factorization, system_matrix; check = false)
+            if issuccess(factorization)
+                scale_by_mass_matrix!(tmp, D)
+                dv .= factorization \ tmp
+            else
+                # The factorization may fail if the time step is too large
+                # and h becomes negative.
+                fill!(dv, NaN)
+            end
+        else
+            factorization = cholesky!(system_matrix)
+            scale_by_mass_matrix!(tmp, D)
+            ldiv!(dv, factorization, tmp)
+        end
+    end
+
+    return nothing
+end
+
+function rhs_sgn_flat_upwind!(dq, q, equations, source_terms, cache)
```

Comment:
Implementing two separate rhs functions for the central and the upwind operators introduces quite some code duplication. Could we maybe make this more elegant? For the other equations I've implemented it in just one function and used branches only for the different cases, where the two versions actually differed. What do you think?

Reply:
Since this would mean quite a few of branches and also dead branches, it is easier to keep it as it is.